### PR TITLE
JAPI-500 String trim flag when reading hpcc files

### PR DESCRIPTION
**Pull request by jpmcmu**
- Added string processing flags / code in BinaryRecordReader
- Implemented unit tests for trim / empty to null conversion

Signed-off-by: James McMullan James.McMullan@lexisnexis.com

## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [X] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).

## Checklist:
- [X] I have created a corresponding JIRA ticket for this submission
- [X] My code follows the code style of this project.
  - [ ] I have applied the Eclipse code-format template provided.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [X] I have read the HPCC Systems CONTRIBUTORS document (https://github.com/hpcc-systems/HPCC-Platform/wiki/Guide-for-contributors).
- [X] The change has been fully tested:
  - [ ] This change does not cause any existing JUnits to fail.
  - [ ] I have include JUnit coverage to test this change
  - [ ] I have performed system test and covered possible regressions and side effects.
- [X] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Premature optimization
  - [ ] This change fixes the problem, not just the symptom

### CODE REVIEW

**Comment by jpmcmu** [comment]
@rpastrana @drealeed please review

**Review by drealeed — APPROVED** [review]

File: dfsclient/src/main/java/org/hpccsystems/dfs/client/BinaryRecordReader.java [hunk 1]
```diff
@@ -149,7 +157,7 @@ public class BinaryRecordReader implements IRecordReader
     private static final int     MASK_32_LOWER_HALF  = 0xffff;
     private static final int     BUFFER_GROW_SIZE    = 8192;
     private static final int     OPTIMIZED_STRING_READ_AHEAD = 32;
-
+
```
Comment (rpastrana) [inline dfsclient/src/main/java/org/hpccsystems/dfs/client/BinaryRecordReader.java:160]: remove

File: dfsclient/src/main/java/org/hpccsystems/dfs/client/BinaryRecordReader.java [hunk 2]
```diff
@@ -131,6 +131,14 @@ public class BinaryRecordReader implements IRecordReader
     private long                 streamPosAfterLastRecord = 0;
     private boolean              isIndex = false;
     private boolean              useDecimalForUnsigned8 = false;
+
```
Comment (rpastrana) [inline dfsclient/src/main/java/org/hpccsystems/dfs/client/BinaryRecordReader.java:134]: let's add a public static final int      NO_STRING_TRIM= 1;

File: dfsclient/src/main/java/org/hpccsystems/dfs/client/BinaryRecordReader.java [hunk 3]
```diff
@@ -914,6 +940,59 @@ private BigDecimal getSignedDecimal(int numDigits, int precision, int dataLen) t
         return ret;
     }
 
+    private void trimStringInScratchBuffer(boolean isUnicode, int[] range)
```
Comment (rpastrana) [inline dfsclient/src/main/java/org/hpccsystems/dfs/client/BinaryRecordReader.java:943]: let's add element level comments, as discussed, see if we can use Character.isWhitespace

File: dfsclient/src/main/java/org/hpccsystems/dfs/client/BinaryRecordReader.java [hunk 4]
```diff
@@ -1239,8 +1332,21 @@ else if ((this.scratchBuffer[strByteLen + bytesScanned] & 0xF8) == 0xF0)
...
+        }
+        
+        strByteLen = strRange[1] - strRange[0];
+        if (strByteLen == 0 && convertEmptyStringsToNull)
+        {
+            return null;
+        }
 
-        return new String(this.scratchBuffer, 0, strByteLen, charset);
+        return new String(scratchBuffer,strRange[0],strByteLen,charset);
```
Comment (rpastrana) [inline dfsclient/src/main/java/org/hpccsystems/dfs/client/BinaryRecordReader.java:1349]: extra white space at end of line

**Review by rpastrana — CHANGES_REQUESTED** [review]
@jpmcmu a few minor comments

**Review by rpastrana — APPROVED** [review]

**Pull request merged**